### PR TITLE
New fallback-dependencies test using Mocha

### PR DESCRIPTION
**Pull request by RailonA**
Created a test script that builds repos using `fallback-dependancies` in the package.json.  Mocha NYC shows 100% test coverage. 

#10

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -22,29 +22,33 @@
...
+      "mocha/no-exclusive-tests": "error"
     }
   },
+  "mocha": {
+    "timeout": "200000ms"
+  },
   "scripts": {
     "codecov": "codecov",
-    "coverage": "c8 --reporter=text --reporter=lcov ava",
+    "coverage": "c8 --reporter=text --reporter=lcov  mocha test/*.js --test",
```
Comment (kethinov) [inline package.json:48]:
```suggestion
    "coverage": "c8 --reporter=text --reporter=lcov mocha test/*.js --test",
```

File: test/test.js [hunk 2]
```diff
@@ -1,5 +1,112 @@
...
+    cleanupTestApp(appDir, (err) => {
+      if (err) {
+        throw err
+      } else {
+        done()
+      }
+    })
+  })
+
+  it('Testing if `fallback-dependencies`  works and creates all files intended', function () {
```
Comment (kethinov) [inline test/test.js:45]:
```suggestion
  it('should create all files intended', function () {
```

File: test/test.js [hunk 3]
```diff
@@ -1,5 +1,112 @@
...
+    assert(fs.existsSync(path.join(__dirname, './clones/repo1/lib/fallback-deps-test-repo-2/lib/fallback-deps-test-repo-3')) === true, './clones/repo1/lib/fallback-deps-test-repo-2/lib/fallback-deps-test-repo-3 does not exist')
+    assert(fs.existsSync(path.join(__dirname, './clones/repo1/lib/fallback-deps-test-repo-2/lib/fallback-deps-test-repo-3/package-lock.json')) === true, './clones/repo1/lib/fallback-deps-test-repo-2/lib/fallback-deps-test-repo-3/package-lock.json does not exist')
+    assert(fs.existsSync(path.join(__dirname, './clones/repo1/lib/fallback-deps-test-repo-2/lib/fallback-deps-test-repo-3/package.json')) === true, './clones/repo1/lib/fallback-deps-test-repo-2/lib/fallback-deps-test-repo-3/package.json does not exist')
+    assert(fs.existsSync(path.join(__dirname, './clones/repo1/lib/fallback-deps-test-repo-2/lib/fallback-deps-test-repo-3/lib')) === false, './clones/repo1/lib/fallback-deps-test-repo-2/lib/fallback-deps-test-repo-3/lib does not exist')
+  })
+
+  it('Testing if `fallback-dependencies` still works if repos is non-presence and replaced with reposFile', function () {
```
Comment (kethinov) [inline test/test.js:56]:
```suggestion
  it('should create intended files using reposFile feature', function () {
```

File: test/test.js [hunk 4]
```diff
@@ -1,5 +1,112 @@
...
+    assert(fs.existsSync(path.join(__dirname, './clones/repo4/lib/fallback-deps-test-repo-5/lib/fallback-deps-test-repo-6')) === true, './clones/repo4/lib/fallback-deps-test-repo-5/lib/fallback-deps-test-repo-6 does exist')
+    assert(fs.existsSync(path.join(__dirname, './clones/repo4/lib/fallback-deps-test-repo-5/lib/fallback-deps-test-repo-6/package-lock.json')) === true, './clones/repo4/lib/fallback-deps-test-repo-5/lib/fallback-deps-test-repo-6/package-lock.json does exist')
+    assert(fs.existsSync(path.join(__dirname, './clones/repo4/lib/fallback-deps-test-repo-5/lib/fallback-deps-test-repo-6/package.json')) === true, './clones/repo4/lib/fallback-deps-test-repo-5/lib/fallback-deps-test-repo-6/package.json does exist')
+    assert(fs.existsSync(path.join(__dirname, './clones/repo4/lib/fallback-deps-test-repo-5/lib/fallback-deps-test-repo-6/lib')) === false, './clones/repo4/lib/fallback-deps-test-repo-5/lib/fallback-deps-test-repo-6/lib does exist')
+  })
+
+  it('Testing if `fallback-dependencies` still works if the repo is not a direct dependency of the root project, append the `:directOnly`', function () {
```
Comment (kethinov) [inline test/test.js:67]:
```suggestion
  it('should create intended files using directOnly feature', function () {
```

File: test/test.js [hunk 5]
```diff
@@ -1,5 +1,112 @@
...
+    require(appendingDirectOnly)()
+    assert(fs.existsSync(path.join(__dirname, './clones/repo7/lib')) === true, './clones/repo7/lib does not exist')
+    assert(fs.existsSync(path.join(__dirname, './clones/repo7/lib/fallback-deps-test-repo-8')) === true, './clones/repo7/lib/fallback-deps-test-repo-8 does not exist')
+    assert(fs.existsSync(path.join(__dirname, './clones/repo7/lib/fallback-deps-test-repo-8/lib')) === true, './clones/repo7/lib/fallback-deps-test-repo-8/lib does not exist')
+  })
+
+  it('Testing if `fallback-dependencies` still works if the repo already exists', function () {
```
Comment (kethinov) [inline test/test.js:74]:
```suggestion
  it('should still work if the repo already exists', function () {
```

File: test/test.js [hunk 6]
```diff
@@ -1,5 +1,112 @@
...
+    require(ifRepoAlreadyExists)()
+    assert(fs.existsSync(path.join(__dirname, './clones/repo10/lib')) === true, './clones/repo10/lib does not exist')
+    assert(fs.existsSync(path.join(__dirname, './clones/repo10/lib/fallback-deps-test-repo-11')) === true, './clones/repo10/lib/fallback-deps-test-repo-11 does not exist')
+  })
+
+  it('Testing if `fallback-dependencies` still works if there is a git pull error!', function () {
```
Comment (kethinov) [inline test/test.js:80]:
```suggestion
  it('should fail gracefully if there is a git pull error', function () {
```

File: test/test.js [hunk 7]
```diff
@@ -1,5 +1,112 @@
...
+  })
+
+  it('Testing if skip installing dependencies for a specific fallback-dependency by using ` -skip-deps` works', function () {
+    require(specificFallbackDependency)()
+    assert(fs.existsSync(path.join(__dirname, './clones/repo16/lib')) === true, './clones/repo16/lib does not exist')
+    assert(fs.existsSync(path.join(__dirname, './clones/repo16/lib/fallback-deps-test-repo-17')) === true, './clones/repo16/lib/fallback-deps-test-repo-17 does not exist')
+    assert(fs.existsSync(path.join(__dirname, './clones/repo16/lib/fallback-deps-test-repo-17/.git')) === true, './clones/repo16/lib/fallback-deps-test-repo-17/.git does not exist')
+  })
+
+  it('Checking to see if  cloneing a specific git tag by using add `-b tag_name` works', function () {
```
Comment (kethinov) [inline test/test.js:93]:
```suggestion
  it('should clone a specific git tag by using add -b tag_name feature', function () {
```

File: test/test.js [hunk 8]
```diff
@@ -1,5 +1,112 @@
...
+    assert(fs.existsSync(path.join(__dirname, './clones/repo19/lib/fallback-deps-test-repo-20/.git')) === true, './clones/repo19/lib/fallback-deps-test-repo-20/.git does not exist')
+  })
+
+  it('Testing if `fallback-dependencies` reclones the repo if the tag does not match the desired -b version number', function () {
+    require(desiredVersion)()
+    assert(fs.existsSync(path.join(__dirname, './clones/repo22/lib')) === true, './clones/repo22/lib does not exist')
+    assert(fs.existsSync(path.join(__dirname, './clones/repo22/lib/fallback-deps-test-repo-23')) === true, './clones/repo22/lib/fallback-deps-test-repo-23 does not exist')
+  })
+
+  it('Testing if `fallback-dependencies` does not update if  repos is not a git repo', function () {
```
Comment (kethinov) [inline test/test.js:106]:
```suggestion
  it('should skip any folder that is not a git repo', function () {
```

**Review by kethinov — CHANGES_REQUESTED** [review]

File: test/test.js [hunk 9]
```diff
@@ -1,5 +1,92 @@
...
+const specificFallbackDependency = path.join(__dirname, './util/specificFallbackDependency.js')
+const cloneingSpecificGitTag = path.join(__dirname, './util/cloneingSpecificGitTag.js')
+const desiredVersion = path.join(__dirname, './util/desiredVersion.js')
+const notAGitRepo = path.join(__dirname, './util/notAGitRepo.js')
+
+const fs = require('fs')
+
+describe('Testing script fallbackDep.js', function () {
+  // delete the test app Directory and start with a clean state after each test
+  after(function (done) {
```
Comment (kethinov) [inline test/test.js:19]: make a `before` for this too

File: test/util/appendingDirectOnly.js [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+    if (fs.existsSync(path.join(process.cwd(), 'node_modules')) === false) {
+      processEnv = process.cwd()
+    } else {
+      processEnv = undefined
+    }
+    appDir = processEnv
+  }
+
+  if (fs.existsSync(`${testSrc}/repos`)) {
+    fs.rmSync(`${testSrc}/repos`, { recursive: true, force: true })
```
Comment (kethinov) [inline test/util/appendingDirectOnly.js:21]: the before/after steps in the tests should take care of this. we can remove all these from the bootstrapping modules

File: test/util/appendingDirectOnly.js [hunk 11]
```diff
@@ -0,0 +1,185 @@
+module.exports = fallbackDependancySandBox
+
+function fallbackDependancySandBox (appDir) {
```
Comment (kethinov) [inline test/util/appendingDirectOnly.js:3]: appDir is never used for anything. we can remove all the logic that depends on it

File: test/util/appendingDirectOnly.js [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+  const fs = require('fs')
+  const path = require('path')
+  const testSrc = path.resolve(__dirname, '../../test')
+  const { execSync } = require('child_process')
+  const repoList = ['repo1', 'repo2', 'repo3']
+
+  if (!appDir) {
+    let processEnv
```
Comment (kethinov) [inline test/util/appendingDirectOnly.js:11]: likewise processEnv is never used for anything. i think we can remove all references to this too

**Review by kethinov — CHANGES_REQUESTED** [review]

File: test/util/appendingDirectOnly.js (hunk 10 above)
Comment (RailonA) [reply test/util/appendingDirectOnly.js]: Had to use this.beforeEach for it to function as expected.

**Review by kethinov — APPROVED** [review]

**Pull request merged**